### PR TITLE
End-matching DID names allowed for listing rules

### LINKED ISSUE #7121: Adding a matching pattern for DIDs in list-rules command
### Description

The list-rules can list rules for a single DID, an entire account or an entire subscription. We would like to add DIDs matching pattern to it. In order to avoid heavy DB query, we suggest below:

`r list-rules --account register_service  raw:Dataset/LSSTCam/202408'*' `

### Motivation

Some of the service accounts have a lot of rules/files and we are only interested to some of them during debugging or a quick checking. Adding a did pattern to limit to the output will avoid the additional filter that the client has to do. In addition, it will help Rucio to reduce the output and speed up.   

### Change

The changes will be in the Client API to allow the did matching pattern and add a DID like clause to the DB query. The changes should be minimum as Rucio already does the most part of this.

### PR DESCRIPTION
**Pull request by voetberg**
Closes #7121

### CODE REVIEW

File: lib/rucio/core/rule.py [hunk 1]
```diff
@@ -1015,6 +1015,14 @@ def list_rules(
                 value = DIDType(value)
             elif key == 'grouping' and isinstance(value, str):
                 value = RuleGrouping(value)
+            elif key == 'name':
+                # First check if it is a real DID or a wildcard
+                if "*" in value:
+                    # Get the matching dids
+                    value = value.replace('*', '%')
+                    stmt = stmt.where(models.ReplicationRule.name.like(value))
+                    continue
```
Comment (mgajek-cern) [inline lib/rucio/core/rule.py:1024]: **Should we support `?` as a single-character wildcard and handle literal `%`/`_` in DIDs?**

Suggested implementation:
```python
elif key == 'name':
    if "*" in value or "?" in value:
       # Prevent overly broad queries
        if value.strip() in ("*", "?", "%", "_"):
            raise ValueError("Pattern too broad - please be more specific")
        # Escape backslashes first to prevent double-escaping
        value = value.replace('\\', '\\\\')
        # Escape SQL wildcards so literal % and _ are treated as normal characters
        value = value.replace('%', r'\%').replace('_', r'\_')
        # Convert shell-style wildcards to SQL wildcards
        value = value.replace('*', '%').replace('?', '_')
        stmt = stmt.where(models.ReplicationRule.name.like(value, escape='\\'))
        continue
```

**Example:**
```python
value = "file?*_%"
```

After transformations: `"file_%\_\%"`

**SQL LIKE interpretation:**
- `file` = literal "file"
- `_` = any single character (from `?`)
- `%` = any sequence (from `*`)
- `\_` = literal underscore (escaped)
- `\%` = literal percent (escaped)

**Result:** Matches strings like `"fileA123_data%"` where `A` is any character and `123` is any sequence.

File: lib/rucio/cli/bin_legacy/rucio.py [hunk 2]
```diff
@@ -1467,22 +1466,26 @@ def list_rules(args, client, logger, console, spinner):
             rules.append(client.get_replication_rule(rule_id))
     elif args.did:
         scope, name = get_scope(args.did, client)
-        meta = client.get_metadata(scope=scope, name=name)
-        rules = client.list_did_rules(scope=scope, name=name)
+        if not name.endswith('*'):
```
Comment (mgajek-cern) [inline lib/rucio/cli/bin_legacy/rucio.py:1469]: This assumes wildcards only appear at the end. What if someone queries `file_*_test` or `*_important`? We might want:

```suggestion
     if "*" not in name:  # No wildcards at all
```

File: tests/test_bin_rucio.py [hunk 3]
```diff
@@ -641,6 +641,23 @@ def test_move_rule_with_arguments(rse_factory, did_factory, rucio_client):
...
+
+    [rule_id] = rucio_client.add_replication_rule(
+        dids=[{'scope': d['did_scope'], 'name': d['did_name']} for d in dataset], account=root_account.external, copies=1, rse_expression=rse_1)
+
+    exitcode, out, err = execute(cmd)
+    assert rule_id in out
+    assert 'ERROR' not in err
+    assert exitcode == 0
+
+
```
Comment (mgajek-cern) [inline tests/test_bin_rucio.py:660]: Should following edge cases be tested as well:

```
    # Test multiple wildcards e.g. "file_*_test_*"
    # Test wildcard at beginning e.g. "*_important"
    # Test wildcard in middle e.g. "file_*_2024"
    # Test no matches e.g. "nonexistent_*"
```

File: lib/rucio/core/rule.py [hunk 4]
```diff
@@ -1015,6 +1015,14 @@ def list_rules(
                 value = DIDType(value)
             elif key == 'grouping' and isinstance(value, str):
                 value = RuleGrouping(value)
+            elif key == 'name':
+                # First check if it is a real DID or a wildcard
+                if "*" in value:
+                    # Get the matching dids
+                    value = value.replace('*', '%')
+                    stmt = stmt.where(models.ReplicationRule.name.like(value))
```
Comment (mgajek-cern) [inline lib/rucio/core/rule.py:1023]: Should we add a `LIMIT` to wildcard queries to reduce excessive resource (like CPU memory, or I/O) consumption for an extended time leading to slow performance or system slowdowns, e.g. 

```suggestion
                    stmt = stmt.where(models.ReplicationRule.name.like(value))
                    stmt = stmt.limit(1000)  # Prevent runaway queries
```

File: lib/rucio/core/rule.py [hunk 5]
```diff
@@ -1015,6 +1015,14 @@ def list_rules(
                 value = DIDType(value)
             elif key == 'grouping' and isinstance(value, str):
                 value = RuleGrouping(value)
+            elif key == 'name':
+                # First check if it is a real DID or a wildcard
```
Comment (mgajek-cern) [inline lib/rucio/core/rule.py:1019]: To be more specific we could instead have following comment:

```suggestion
               # Check if name contains wildcard pattern and convert to SQL LIKE syntax
```

File: lib/rucio/core/rule.py [hunk 6]
```diff
@@ -1015,6 +1015,14 @@ def list_rules(
...
+                if "*" in value:
```
Comment (mgajek-cern) [inline lib/rucio/core/rule.py:1020]: `LIKE` queries with leading wildcards (%something) can't use indexes efficiently, e.g.

```sql
-- Assume an indexed column `name`

-- 1. Wildcard at the end (can use index efficiently)
SELECT * FROM files WHERE name LIKE 'report%';

-- 2. Wildcard at the beginning (index usually not used, slower)
SELECT * FROM files WHERE name LIKE '%report';
``` 


We could add following comment:

```
# Note: Wildcards at the beginning of the pattern (*something) 
# may result in slower queries as they cannot use database indexes
```

**Review by mgajek-cern — COMMENTED** [review]
**NOTE:** Reviewing this PR based on the understanding that wildcards can appear anywhere (in the beginning, middle, or end) according to the issue.

Here are a few remarks/questions/findings:

**Comment by voetberg** [comment]
> **NOTE:** Reviewing this PR based on the understanding that wildcards can appear anywhere (in the beginning, middle, or end) according to the issue.
> 
> Here are a few remarks/questions/findings:

Sorry for the misleading issue title. I'm using "wildcard" as a shorthand for "match the end of a DID pattern" - as the issue suggests.

**Comment by mgajek-cern** [comment]
> Sorry for the misleading issue title. I'm using "wildcard" as a shorthand for "match the end of a DID pattern" - as the issue suggests.

Thanks for clarifying. Before someone else continues with the review of the simpler end-matching implementation, can you confirm this scope has been agreed upon with the issue creator (@yuyiguo)? 

Just want to make sure we're all aligned on the requirements (end-matching only vs. full wildcard support) before the next reviewer dives in.

**Comment by voetberg** [comment]
> > Sorry for the misleading issue title. I'm using "wildcard" as a shorthand for "match the end of a DID pattern" - as the issue suggests.
> 
> Thanks for clarifying. Before someone else continues with the review of the simpler end-matching implementation, can you confirm this scope has been agreed upon with the issue creator (@yuyiguo)?
> 
> Just want to make sure we're all aligned on the requirements (end-matching only vs. full wildcard support) before the next reviewer dives in.

I talked with Yuyi IRL and we agreed that end-matching is the better option. I'll change the name of the PR appropriately

**Comment by bari12** [comment]
I think this needs to support wildcards anywhere in the name, not just the end. 
I understand that Yuyis use-case does not require it, but if we add this functionality, the assumption will be that it works the same as a `rucio did list` - where we do allow it at any point in the pattern (except the beginning as @mgajek-cern pointed out) - Dealing with this differently in `rucio rule list` is a mistake I think, it should be equivalent.

File: lib/rucio/core/rule.py (hunk 1 above)
Comment (bziemons) [reply lib/rucio/core/rule.py:1024]: this would have been a discussion for the issue, but since this is already the PR I see your comment as out of scope. Open a new issue if you really need this feature.
